velox/type/tests/DecimalTest.cpp

```diff
@@ -287,84 +291,202 @@ TEST(DecimalAggregateTest, adjustSumForOverflow) {
-  assertRescaleDouble<int128_t>(
+  assertRescaleFloatingPoint<double, int128_t>(
       0.999999999999999, DECIMAL(38, 18), 999'999'999'999'999'000);
-  assertRescaleDouble<int128_t>(
+  assertRescaleFloatingPoint<double, int128_t>(
       0.123456789123123, DECIMAL(38, 18), 123'456'789'123'123'000);
-  assertRescaleDouble<int64_t>(21.54551, DECIMAL(12, 3), 21546);
+  assertRescaleFloatingPoint<double, int64_t>(21.54551, DECIMAL(12, 3), 21546);
 
-  assertRescaleDoubleFail<int128_t>(
+  assertRescaleFloatingPointFail<double, int128_t>(
       std::numeric_limits<double>::max(), DECIMAL(38, 38), "Result overflows.");
-  assertRescaleDouble<int128_t>(
+  assertRescaleFloatingPoint<double, int128_t>(
       std::numeric_limits<double>::min(), DECIMAL(38, 2), 0);
-  assertRescaleDoubleFail<int128_t>(
+  assertRescaleFloatingPointFail<double, int128_t>(
       std::numeric_limits<double>::lowest(),
       DECIMAL(38, 2),
       "Result overflows.");
 
-  assertRescaleDoubleFail<int64_t>(
+  assertRescaleFloatingPointFail<double, int64_t>(
       NAN, DECIMAL(10, 2), "The input value should be finite.");
-  assertRescaleDoubleFail<int64_t>(
+  assertRescaleFloatingPointFail<double, int64_t>(
       INFINITY, DECIMAL(10, 2), "The input value should be finite.");
 
-  assertRescaleDoubleFail<int64_t>(
+  assertRescaleFloatingPointFail<double, int64_t>(
       9999999999999999999999.99, DECIMAL(10, 2), "Result overflows.");
-  assertRescaleDoubleFail<int64_t>(
+  assertRescaleFloatingPointFail<double, int64_t>(
       static_cast<double>(
           static_cast<int128_t>(std::numeric_limits<int64_t>::max()) + 1),
       DECIMAL(10, 2),
       "Result overflows.");
-  assertRescaleDoubleFail<int64_t>(
+  assertRescaleFloatingPointFail<double, int64_t>(
       static_cast<double>(
           static_cast<int128_t>(std::numeric_limits<int64_t>::min()) - 1),
       DECIMAL(10, 2),
       "Result overflows.");
-  assertRescaleDoubleFail<int64_t>(
+  assertRescaleFloatingPointFail<double, int64_t>(
       static_cast<double>(DecimalUtil::kShortDecimalMax),
       DECIMAL(10, 2),
       "Result overflows.");
-  assertRescaleDoubleFail<int64_t>(
+  assertRescaleFloatingPointFail<double, int64_t>(
       static_cast<double>(DecimalUtil::kShortDecimalMin),
       DECIMAL(10, 2),
       "Result overflows.");
-  assertRescaleDoubleFail<int128_t>(
+  assertRescaleFloatingPointFail<double, int128_t>(
       static_cast<double>(DecimalUtil::kLongDecimalMax),
       DECIMAL(20, 2),
       "Result overflows.");
-  assertRescaleDoubleFail<int128_t>(
+  assertRescaleFloatingPointFail<double, int128_t>(
       static_cast<double>(DecimalUtil::kLongDecimalMin),
       DECIMAL(20, 2),
       "Result overflows.");
 
-  assertRescaleDoubleFail<int64_t>(
+  assertRescaleFloatingPointFail<double, int64_t>(
       99999.99, DECIMAL(6, 4), "Result cannot fit in the given precision 6.");
 }
 
+TEST(DecimalTest, rescaleReal) {
+  assertRescaleFloatingPoint<float, int64_t>(
```

Comment:
Using HugeInt::parse helps readability. Thanks. 

I wonder if we can improve readability a bit more by introducing some helper functions.

```
assertRescaleReal(-3333.03, DECIMAL(10, 4), -33'330'300);
assertRescaleReal(-3333.03, DECIMAL(20, 1), -33'330);

assertRescaleReal(-3333.03, DECIMAL(38, 18), "-333303" + zeros(16));
```

Also, wondering if it makes sense to add some cases where we do loose precision.

Reply:
Added assertRescaleReal, assertRescaleDouble and zeros methods. The precision loss cases mentioned in this PR are added in tests.